packages/ckeditor5-utils/src/dom/position.ts

```diff
@@ -99,12 +100,44 @@ export function getOptimalPosition( { element, target, positions, limiter, fitIn
 
 	let bestPosition: Position;
 
+	// Get intersection of all scrollable ancestors of `target`.
+	const allScrollableAncestors = getScrollableAncestors( target as HTMLElement );
+	const ancestorsIntersection: Rect | null = allScrollableAncestors
+		.reverse()
+		.map( item => new Rect( item as HTMLElement ) )
+		.reduce( ( ancestorIntersectionAccumulator, ancestorRect ): Rect => {
+			if ( !ancestorIntersectionAccumulator || ancestorIntersectionAccumulator.top === undefined ) {
+				return ancestorRect;
+			}
+
+			if ( !ancestorRect || ancestorRect.top === undefined ) {
+				return ancestorIntersectionAccumulator;
+			}
+
+			return ancestorIntersectionAccumulator.getIntersection( ancestorRect ) as Rect;
+		} );
+
 	// @if CK_DEBUG_POSITION // RectDrawer.clear();
 	// @if CK_DEBUG_POSITION // RectDrawer.draw( targetRect, { outlineWidth: '5px' }, 'Target' );
 
 	const viewportRect = fitInViewport && getConstrainedViewportRect( viewportOffsetConfig ) || null;
 	const positionOptions = { targetRect, elementRect, positionedElementAncestor, viewportRect };
 
+	if ( ( !ancestorsIntersection ) || ( !targetRect.getVisible() ) ) {
+		// Default last position is `viewportHidden`.
```

Comment:
You don't know that. `getOptimalPosition()` gets an arbitrary set of positions. It is not used by balloon system only but also by dropdowns to position their panels.